src/transport/tests/TestGroupMessageCounter.cpp

```diff
@@ -0,0 +1,470 @@
+
+    i = 1;
+    do
+    {
+        err = mGroupPeerMsgCounter.FindOrAddPeer(++fabricIndex, peerNodeId, false, counter);
+        i++;
+    } while (err != CHIP_ERROR_TOO_MANY_PEER_NODES);
+    NL_TEST_ASSERT(inSuite, i == CHIP_CONFIG_MAX_FABRICS + 1);
+}
+
+void RemovePeerTest(nlTestSuite * inSuite, void * inContext)
+{
+    NodeId peerNodeId                             = 1234;
+    FabricIndex fabricIndex                       = 1;
+    CHIP_ERROR err                                = CHIP_NO_ERROR;
+    chip::Transport::PeerMessageCounter * counter = nullptr;
+    chip::Transport::GroupPeerTable mGroupPeerMsgCounter;
+
+    // Fill table up (max fabric and mac peer)
+    for (uint32_t it = 0; it < CHIP_CONFIG_MAX_FABRICS; it++)
+    {
+        for (uint32_t peerId = 0; peerId < CHIP_CONFIG_MAX_GROUP_CONTROL_PEERS; peerId++)
+        {
+            err = mGroupPeerMsgCounter.FindOrAddPeer(fabricIndex, peerNodeId++, true, counter);
+        }
+        fabricIndex++;
+    }
+    // Verify that table is indeed full (for control Peer)
+    err = mGroupPeerMsgCounter.FindOrAddPeer(99, 99, true, counter);
+    NL_TEST_ASSERT(inSuite, err == CHIP_ERROR_TOO_MANY_PEER_NODES);
+
+    // Clear all Peer
+    fabricIndex = 1;
+    peerNodeId  = 1234;
+    for (uint32_t it = 0; it < CHIP_CONFIG_MAX_FABRICS; it++)
+    {
+        for (uint32_t peerId = 0; peerId < CHIP_CONFIG_MAX_GROUP_CONTROL_PEERS; peerId++)
+        {
+            err = mGroupPeerMsgCounter.RemovePeer(fabricIndex, peerNodeId++, true);
+            NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+        }
+        fabricIndex++;
+    }
+
+    // Try re-adding the previous peer without any error
+    err = mGroupPeerMsgCounter.FindOrAddPeer(99, 99, true, counter);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+}
+
+void PeerRetrievalTest(nlTestSuite * inSuite, void * inContext)
+{
+    NodeId peerNodeId                              = 1234;
+    FabricIndex fabricIndex                        = 1;
+    CHIP_ERROR err                                 = CHIP_NO_ERROR;
+    chip::Transport::PeerMessageCounter * counter  = nullptr;
+    chip::Transport::PeerMessageCounter * counter2 = nullptr;
+    chip::Transport::GroupPeerTable mGroupPeerMsgCounter;
+
+    err = mGroupPeerMsgCounter.FindOrAddPeer(fabricIndex, peerNodeId, true, counter);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, counter != nullptr);
+
+    err = mGroupPeerMsgCounter.FindOrAddPeer(99, 99, true, counter2);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, counter2 != nullptr);
+    NL_TEST_ASSERT(inSuite, counter2 != counter);
+
+    err = mGroupPeerMsgCounter.FindOrAddPeer(fabricIndex, peerNodeId, true, counter2);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, counter2 == counter);
+}
+
+void CounterCommitRolloverTest(nlTestSuite * inSuite, void * inContext)
```

Comment:
Here is the minimal set of tests I think we need.  For all of these, the assumption is that we have no synced state, then a counter value N comes in, then some set of tests happens.

1) Repeat, starting with above "just got a value of N" state each time, for all k such that `1 <= k <= 2 * CHIP_CONFIG_MESSAGE_COUNTER_WINDOW_SIZE`:
    a) A counter value of `N + k` comes in, we detect it as valid and commit it.
    b) A counter value of `N` comes in, we detect it as duplicate.
    c) A counter value between `N - CHIP_CONFIG_MESSAGE_COUNTER_WINDOW` and `N + k - CHIP_CONFIG_MESSAGE_COUNTER_WINDOW` (but _not_ including `N + k - CHIP_CONFIG_MESSAGE_COUNTER_WINDOW`) comes in, we treat it as duplicate.
    d) A counter value of `N + k - CHIP_CONFIG_MESSAGE_COUNTER_WINDOW` comes in, is treated as valid.  This test will presumably fail for now and can be left #if 0.

   This test should be performed for at least `N = 0`, `N = 10`, `N = 0x7FFFFFFF`, `N = 0x80000000`, `N = 0x80000001`, `N = 0x80000002`, `N = 0xFFFFFFF0`, `N = 0xFFFFFFFF`.

2) Again starting with a "just got N" counter, but now the substeps are all done without creating a new counter until we change N:
    a) Some set of values `N - k` come in, for `0 < k < CHIP_CONFIG_MESSAGE_COUNTER_WINDOW`.  All of those should be considered valid and committed.
    b) Counter value N + 3 comes in.
    c) The same set of value `N - k` come in as in step (a) and are all considered duplicates.
    d) The values that were not in the set in step (a) (but are at least `N + 3 - CHIP_CONFIG_MESSAGE_COUNTER_WINDOW`) come in, and all are treated as allowed.

    This test should also be peformed for at least `N = 0`, `N = 10`, `N = 0x7FFFFFFF`, `N = 0x80000000`, `N = 0xFFFFFFF0`, `N = 0xFFFFFFFF`.

3) Again starting with a "just got N" state each time, repeat for `2^31 - 5 <= k <= 2^31 - 1` the steps from test 1.  Do this for all the `N` in that test.

4) Again starting with a "just got N" state each time, repeat for `2^31 <= k <= 2^31 + 2`: a new counter with the value `N + k` comes in and is treated as a duplicate.  Do this for all of our `N` values that we do the other tests for.

This should be doable with a loop over an array of `N` values, and for each one runnning these 4 tests.  Some of which will then loop over `k` values or arrays of `k` values.

I'm OK with these tests happening in a followup as desired.

There might be more cases that could use testing too, but I think the above should hit most of the various situations where we might or might not hit rollover, etc.

Reply:
Will do with #15214